services/brig/src/Brig/Data/Blacklist.hs

```diff
@@ -27,3 +39,33 @@ keySelect = "SELECT key FROM blacklist WHERE key = ?"
 
 keyDelete :: PrepQuery W (Identity Text) ()
 keyDelete = "DELETE FROM blacklist WHERE key = ?"
+
+--------------------------------------------------------------------------------
+-- Excluded phone prefixes
+
+insertPrefix :: MonadClient m => PhonePrefix -> m ()
+insertPrefix prefix = retry x5 $ write prefixInsert (params Quorum (Identity $ fromPhonePrefix prefix))
+
+deletePrefix :: MonadClient m => PhonePrefix -> m ()
+deletePrefix prefix = retry x5 $ write prefixDelete (params Quorum (Identity $ fromPhonePrefix prefix))
+
+getAllPrefixes :: MonadClient m => PhonePrefix -> m [PhonePrefix]
+getAllPrefixes prefix = do
+    let prefixes = fromPhonePrefix <$> allPrefixes (fromPhonePrefix prefix)
+    results <- fmap runIdentity <$> retry x1 (query prefixSelectAll (params Quorum (Identity $ prefixes)))
+    return (PhonePrefix <$> results)
+
+existsAnyPrefix :: MonadClient m => Phone -> m Bool
+existsAnyPrefix phone = do
+    let prefixes = fromPhonePrefix <$> allPrefixes (fromPhone phone)
+    results <- fmap runIdentity <$> retry x1 (query prefixSelectAll (params Quorum (Identity $ prefixes)))
+    return $ (not . null) results
+
+prefixInsert :: PrepQuery W (Identity Text) ()
+prefixInsert = "INSERT INTO excluded_phones (prefix) VALUES (?)"
```

Comment:
i've seen this pattern of flipping two words in a composite to make up a new identifier, where the difference in the names has nothing to do with the difference in semantics.  didn't like it back then either, but consistency may be more important.

also, i'm not sure the two identifiers need to be top-level, as each `PrepQuery` is called exactly once (no?).

Reply:
I can make the CQL a sub function, [if you prefer?](https://github.com/wireapp/wire-server/pull/593/commits/69e22b6cb0b1f4339239d7342e9a1577b0ffeffa)